odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+    dx : `float`, optional
+        Scalars specifying the sample distances in each dimension `axis`.
+        Default distance: 1.
+    edge_order : {1, 2}, optional
+        Partial derivative is calculated using Nth order accurate
+        differences at the boundaries. Default edge order: 2
+    zero_padding : `bool`, optional
+        Implicit zero padding. Assumes values outside the domain of f to be
+        zero. Default: False
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Returns an N-dimensional array of the same shape as `f`
+
+    Examples
+    --------
+    >>> f = np.arange(10, dtype=float)
+    >>> discrete_part_deriv(f)
+    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
+    >>> discrete_part_deriv(f, dx=0.5)
+    array([ 2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.])
+    >>> discrete_part_deriv(f, dx=1.0, zero_padding=True)
+    array([ 0.5,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. , -4. ])
+    >>> df1 = discrete_part_deriv(np.sin(f/10*np.pi), edge_order=1)
+    >>> df2 = discrete_part_deriv(np.sin(f/10*np.pi), edge_order=2)
+    >>> np.array_equal(df1[1:-1], df2[1:-1])
+    True
+    >>> df1[0] == df2[0]
+    False
+    >>> df1[-1] == df2[-1]
+    False
+    >>> n = 5
+    >>> f = np.arange(n, dtype=float)
+    >>> f = f * f.reshape((n,1))
+    >>> discrete_part_deriv(f, 0)
+    array([[-0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.]])
+    >>> discrete_part_deriv(f, 1)
+    array([[-0.,  0.,  0.,  0.,  0.],
+           [ 1.,  1.,  1.,  1.,  1.],
+           [ 2.,  2.,  2.,  2.,  2.],
+           [ 3.,  3.,  3.,  3.,  3.],
+           [ 4.,  4.,  4.,  4.,  4.]])
+    >>> try:
+    ...    discrete_part_deriv(f, 2)
+    ... except IndexError as e:
+    ...    print(e)
+    Axis paramater (2) exceeds number of dimensions (2).
+    """
+
+    f_data = np.asanyarray(f)
+    ndim = f_data.ndim
+
+    # create slice objects --- initially all are [:, :, ..., :]
+    # noinspection PyTypeChecker
+    slice1 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice2 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice3 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice4 = [slice(None)] * ndim
+
+    try:
+        if f_data.shape[axis] < 2:
+            raise ValueError("Shape of array too small to calculate a "
+                             "numerical gradient, at least two elements are "
+                             "required.")
+    except IndexError:
+        raise IndexError("Axis paramater ({0}) exceeds number of dimensions "
+                         "({1}).".format(axis, ndim))
+
+    out = np.empty_like(f_data)
+
+    # Numerical differentiation: 2nd order interior
+    slice1[axis] = slice(1, -1)
+    slice2[axis] = slice(2, None)
+    # noinspection PyTypeChecker
```

Comment:
We shouldnt riddle our code with these, it's not used by all tools and makes it harder to read. Imo, remove them.